File: device-sdk/src/main/java/io/oddworks/device/metric/OddMetric.java
package io.oddworks.device.metric;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import io.oddworks.device.service.OddBus;

public abstract class OddMetric implements OddBus.OddRxBusEvent {
    private static final String TAG = OddMetric.class.getSimpleName();
    private static final String TYPE = "event";
    public static final String ACTION_APP_INIT = "app:init";
    public static final String ACTION_VIEW_LOAD = "view:load";
    public static final String ACTION_VIDEO_PLAY = "video:play";
    public static final String ACTION_VIDEO_PLAYING = "video:playing";
    public static final String ACTION_VIDEO_STOP = "video:stop";
    public static final String ACTION_VIDEO_ERROR = "video:error";

    protected String organizationId;
    protected String contentType;
    protected String contentId;


    public abstract String getAction();
    public abstract boolean getEnabled();

    public String getType() {
        return TYPE;
    }

    public String getOrganizationId() {
        return organizationId;
    }

    public String getContentType() {
        return contentType;
    }

    public String getContentId() {
        return contentId;
    }

    public OddMetric setOrganizationId(String organizationId) {
        this.organizationId = organizationId;
        return this;
    }

    public OddMetric setContentType(String contentType) {
        this.contentType = contentType;
        return this;
    }

    public OddMetric setContentId(String contentId) {
        this.contentId = contentId;
        return this;
    }

    public JSONObject toJSONObject() {
        JSONObject json = new JSONObject();
        JSONObject attributes = new JSONObject();

        try {
            json.put("type", getType());

            attributes.put("organizationId", getOrganizationId());
            attributes.put("action", getAction());
            attributes.put("contentType", getContentType());
            attributes.put("contentId", getContentId());

            json.put("attributes", attributes);
        } catch (JSONException e) {
            Log.d(TAG, e.toString());
        }

        return json;
    }

    @Override
    public String toString() {
        return TAG + "(" +
                "type=" + getType() + ", " +
                "organizationId=" + getOrganizationId() + ", " +
                "action=" + getAction() + ", " +
                "contentType=" + getContentType() + ", " +
                "contentId=" + getContentId() +
                ")";

    }
}

File: device-sdk/src/main/java/io/oddworks/device/metric/OddMetric.kt
package io.oddworks.device.metric

import android.util.Log

import org.json.JSONException
import org.json.JSONObject

import io.oddworks.device.service.OddRxBus

abstract class OddMetric(protected val contentType: String?, protected val contentId: String?, protected val meta: JSONObject?) : OddRxBus.OddRxBusEvent {
    abstract val action: String
    abstract val enabled: Boolean

    val type: String
        get() = TYPE

    open fun toJSONObject(): JSONObject {
        val json = JSONObject()
        val attributes = JSONObject()

        try {
            val data = JSONObject()
            data.put("type", type)

            attributes.put("action", action)
            attributes.put("contentType", contentType)
            attributes.put("contentId", contentId)

            data.put("attributes", attributes)
            data.put("meta", meta)
            json.put("data", data);
        } catch (e: JSONException) {
            Log.d(TAG, e.toString())
        }

        return json
    }

    override fun toString(): String {
        return "$TAG (type=$type, action=$action, contentType=$contentType, contentId=$contentId, meta=${meta.toString()})"
    }

    companion object {
        private val TAG = OddMetric::class.java.simpleName
        private val TYPE = "event"

        val ACTION_APP_INIT = "app:init"
        val ACTION_VIEW_LOAD = "view:load"
        val ACTION_VIDEO_PLAY = "video:play"
        val ACTION_VIDEO_PLAYING = "video:playing"
        val ACTION_VIDEO_STOP = "video:stop"
        val ACTION_VIDEO_ERROR = "video:error"
        val ACTION_VIDEO_LOAD = "video:load"
        val ACTION_USER_NEW = "user:new"
    }
}